runtime-modules/storage/src/lib.rs

```diff
@@ -3942,4 +3738,336 @@ impl<T: Trait> Module<T> {
+                        }
+                        BagOperationParams::<T>::Delete => {
+                            bk.unregister_bag_assignment();
+                            (id.clone(), bk)
+                        }
+                        _ => (id.clone(), bk), // no op in case of objects upload/delete operation
+                    })
+                    .ok()
+            })
+            .collect()
+    }
+
+    fn compute_net_prize(
+        init_value: NetDeletionPrize<T>,
+        num_obj_to_create: usize,
+        num_obj_to_delete: usize,
+    ) -> NetDeletionPrize<T> {
+        let amnt = T::DataObjectDeletionPrize::get();
+        init_value
+            .add_balance(
+                iter::repeat(amnt)
+                    .take(num_obj_to_create)
+                    .fold(BalanceOf::<T>::zero(), |acc, x| acc.saturating_add(x)),
+            )
+            .sub_balance(
+                iter::repeat(amnt)
+                    .take(num_obj_to_delete)
+                    .fold(BalanceOf::<T>::zero(), |acc, x| acc.saturating_add(x)),
+            )
+    }
+
+    /// Utility function that checks existence for bag deletion / update &
+    /// verifies that dynamic bag does not exists in case of bag creation
+    /// returning a new one
+    fn retrieve_dynamic_bag(bag_op: &BagOperation<T>) -> Result<Bag<T>, DispatchError> {
+        if let BagOperationParams::<T>::Create(deletion_prize, _) = bag_op.params {
+            let dyn_bag_id = bag_op.bag_id.clone().ensure_is_dynamic_bag::<T>()?;
+
+            Self::ensure_bag_exists(&bag_op.bag_id).map_or_else(
+                |_| {
+                    Ok(Bag::<T>::default()
+                        .with_prize(deletion_prize)
+                        .with_storage_buckets(StorageBucketPicker::<T>::pick_storage_buckets(
+                            dyn_bag_id.clone().into(),
+                        ))
+                        .with_distribution_buckets(
+                            DistributionBucketPicker::<T>::pick_distribution_buckets(
+                                dyn_bag_id.into(),
+                            ),
+                        ))
+                },
+                |_| Err(Error::<T>::DynamicBagExists.into()),
+            )
+        } else {
+            Self::ensure_bag_exists(&bag_op.bag_id)
+        }
+    }
+
+    /// Utility function that does the bulk of bag / data objects operation
+    fn try_mutating_storage_state(
+        account_id: T::AccountId,
+        bag_op: BagOperation<T>,
+    ) -> DispatchResult {
+        // attempt retrieve existing or create new one if operation is create
+        let Bag::<T> {
+            stored_by,
+            distributed_by,
+            deletion_prize,
+            objects_total_size,
+            objects_number,
+        } = Self::retrieve_dynamic_bag(&bag_op)?;
+
+        let bag_id_variant = bag_op.ensure_bag_id_valid_for_op::<T>()?;
+        let dyn_bag_id = bag_id_variant.ensure_is_dynamic_bag::<T>().ok();
```

Comment:
Why did you change the previous line with `?` ?
`let dyn_bag_id = bag_id_variant.ensure_is_dynamic_bag::<T>()?;`

You added a comment about `Some(..)` later which is avoidable.

Reply:
I cannot address the comment: 
If I call `bag_id_variant.ensure_is_dynamic_bag` then it throws errors when you try to update assets on static bags

Reply:
This is really the best I can do

Reply:
I tried a different solution since:
``` rust
         let bag_id_variant = bag_op.bag_id.ensure_is_dynamic_bag::<T>()?; 
```
Makes the function fail for static bags updates